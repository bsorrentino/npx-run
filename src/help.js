import splitArgs from 'string-argv';

import {print} from './strings';
import {formattedScriptName, formattedScript} from './colorize';


export const scriptCompare = (name1, name2)=> {
  name1 = name1 === 'default' ? '' : name1;
  name2 = name2 === 'default' ? '' : name2;

  if (name1 > name2) {
    return 1;
  } else if (name1 < name2) {
    return -1;
  }
  return 0;
};


const printScripts = (scripts)=> {
  const scriptNames = Object.keys(scripts).sort(scriptCompare);
  const maxLen = Math.max(...scriptNames.map((key)=> key.length));

  print`{underline scripts}:`;

  for (const scriptName of scriptNames) {
    const name = formattedScriptName(scriptName);
    const command = splitArgs(scripts[scriptName]);
    const preview = formattedScript(command, scripts);
    const padding = maxLen - scriptName.length;

    print`  ${name} ${' '.repeat(padding)} ${preview}`;
  }

  return scriptNames.length;
};


/* eslint-disable max-len */
export const showHelp = (scripts)=> {
  print`
    npx {bold run} {dim [run-args] [node-args]} <{green script}|{blue pattern}> [srcipt-args] ...

    {underline Examples}:
      npx {bold run}
      npx {bold run} {green clean} {green build}
      npx {bold run} {green config} --env='testing' {green build} {green deploy}
      npx {bold run} {blue lint:*}
      npx {bold run} {dim --inspect-brk} {green jest}

    {underline default script}:
      When calling without any arguemtns 'npx run' will run the
      {bold.green default} script if it has been defined.

    {underline run-args}:
      --help         Show this help.
      --dry-run      List scripts to be run, but don't run them.

    {underline node-args}:
      Any args before the first script are sent to npx as node args.

      --inspect      Run node with --inspect to debug scripts.
      --inspect-brk  Run node with --inspect-brk to debug scripts
  `;

  if (!printScripts(scripts)) {
    print`
      {red no scripts discoverd}

      You can add scripts using:

        ./package.json   - using npm run-scripts
        ./scripts module - exporting a single default object working
                           the same as the package.json run-scripts.
        ./scripts/*.js   - using any loadable nodejs module which is run
                           as a script given its basename

      ./scripts/*.js modules will override scripts of the same name defined in
      the ./scripts module.
      Scripts defined in the ./scripts module will override scripts of the same
      name defined in ./package.json.
    `;
  }

};
