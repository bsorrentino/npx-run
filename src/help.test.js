import {showHelp} from './help';


let mockStdout = '';

jest.mock('./globals', ()=> ({
  process: {
    stdout: {
      write(data) {
        mockStdout += data;
      }
    }
  }
}));


describe('showHelp', ()=> {
  beforeEach(()=> {
    mockStdout = '';
  });


  it('prints help for missing scripts', ()=> {
    const scripts = {};

    showHelp(scripts);

    expect(mockStdout).toMatchSnapshot();
  });


  it('prints help with list of scripts', ()=> {
    const scripts = {
      test: 'echo "Works on my machine!"'
    };

    showHelp(scripts);

    expect(mockStdout).toMatchSnapshot();
  });
});
