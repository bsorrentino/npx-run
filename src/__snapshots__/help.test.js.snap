// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`showHelp prints help for missing scripts 1`] = `
"npx [1mrun[22m [2m[run-args] [node-args][22m <[32mscript[39m|[34mpattern[39m> [srcipt-args] ...

[4mExamples[24m:
  npx [1mrun[22m
  npx [1mrun[22m [32mclean[39m [32mbuild[39m
  npx [1mrun[22m [32mconfig[39m --env='testing' [32mbuild[39m [32mdeploy[39m
  npx [1mrun[22m [34mlint:*[39m
  npx [1mrun[22m [2m--inspect-brk[22m [32mjest[39m

[4mdefault script[24m:
  When calling without any arguemtns 'npx run' will run the
  [1m[32mdefault[39m[22m script if it has been defined.

[4mrun-args[24m:
  --help         Show this help.
  --dry-run      List scripts to be run, but don't run them.

[4mnode-args[24m:
  Any args before the first script are sent to npx as node args.

  --inspect      Run node with --inspect to debug scripts.
  --inspect-brk  Run node with --inspect-brk to debug scripts

[4mscripts[24m:
[31mno scripts discoverd[39m

You can add scripts using:

  ./package.json   - using npm run-scripts
  ./scripts module - exporting a single default object working
                     the same as the package.json run-scripts.
  ./scripts/*.js   - using any loadable nodejs module which is run
                     as a script given its basename

./scripts/*.js modules will override scripts of the same name defined in
the ./scripts module.
Scripts defined in the ./scripts module will override scripts of the same
name defined in ./package.json.

"
`;

exports[`showHelp prints help with list of scripts 1`] = `
"npx [1mrun[22m [2m[run-args] [node-args][22m <[32mscript[39m|[34mpattern[39m> [srcipt-args] ...

[4mExamples[24m:
  npx [1mrun[22m
  npx [1mrun[22m [32mclean[39m [32mbuild[39m
  npx [1mrun[22m [32mconfig[39m --env='testing' [32mbuild[39m [32mdeploy[39m
  npx [1mrun[22m [34mlint:*[39m
  npx [1mrun[22m [2m--inspect-brk[22m [32mjest[39m

[4mdefault script[24m:
  When calling without any arguemtns 'npx run' will run the
  [1m[32mdefault[39m[22m script if it has been defined.

[4mrun-args[24m:
  --help         Show this help.
  --dry-run      List scripts to be run, but don't run them.

[4mnode-args[24m:
  Any args before the first script are sent to npx as node args.

  --inspect      Run node with --inspect to debug scripts.
  --inspect-brk  Run node with --inspect-brk to debug scripts

[4mscripts[24m:
  [1m[32mdefault[39m[22m          [1mrun[22m [32mclean[39m [32mtest[39m
  [32mclean[39m            rimraf [2m./build[22m
  [32mcreate-pkg-json[39m  echo [2mdone[22m
  [32mdry[39m              [1mrun[22m [2m--dry-run[22m [32mclean[39m --verbose
  [32mtest[39m             echo [2mWorks on my machine![22m
"
`;
